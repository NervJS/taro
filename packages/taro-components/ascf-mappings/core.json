{
  "Button": {
    "size": {
      "ascfProp": "dimension",
      "valueMap": {
        "mini": "XS",
        "small": "S",
        "default": "M",
        "large": "L"
      },
      "required": false,
      "default": "M",
      "docLink": "https://ascf.docs/components/button#dimension"
    },
    "loading": {
      "ascfProp": "isLoading",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/button#isloading"
    }
  },
  "Input": {
    "password": {
      "ascfProp": "secretInput",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/input#secretinput"
    }
  }
}
