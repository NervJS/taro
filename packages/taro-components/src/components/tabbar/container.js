import 'weui'
import Nerv from 'nervjs'
import classNames from 'classnames'
import View from '../view'
export default class TabbarContainer extends Nerv.Component {
  render () {
    const { children, className, ...reset } = this.props
    const cls = classNames('taro-tabbar__container', className)
    return (
      <View className={cls} {...reset}>
        {children}
      </View>
    )
  }
}
