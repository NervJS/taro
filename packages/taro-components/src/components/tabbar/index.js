import 'weui'
import Taro from '@tarojs/taro-h5'
import Nerv, { findDOMNode } from 'nervjs'
import classNames from 'classnames'
import resolvePathname from 'resolve-pathname'

import TabbarItem from './tabbarItem'
import { splitUrl } from '../../utils'
import './style/index.scss'

// const removeLeadingSlash = str => str.replace(/^\.?\//, '')
// const removeTrailingSearch = str => str.replace(/\?[\s\S]*$/, '')
const addLeadingSlash = str => str[0] === '/' ? str : `/${str}`

const hasBasename = (path, prefix) =>
  new RegExp('^' + prefix + '(\\/|\\?|#|$)', 'i').test(path)

const stripBasename = (path, prefix) =>
  hasBasename(path, prefix) ? path.substr(prefix.length) : path

const STATUS_SHOW = 0
const STATUS_HIDE = 1
const STATUS_SLIDEOUT = 2

const basicTabBarClassName = 'taro-tabbar__tabbar'
const hideTabBarClassName = 'taro-tabbar__tabbar-hide'
const hideTabBarWithAnimationClassName = 'taro-tabbar__tabbar-slideout'

class Tabbar extends Nerv.Component {
  constructor (props) {
    super(...arguments)
    const list = props.conf.list
    const customRoutes = props.conf.customRoutes
    if (
      Object.prototype.toString.call(list) !== '[object Array]' ||
      list.length < 2 ||
      list.length > 5
    ) {
      throw new Error('tabBar 配置错误')
    }

    this.homePage = addLeadingSlash(props.homePage)
    this.customRoutes = []
    for (let key in customRoutes) {
      this.customRoutes.push([key, customRoutes[key]])
    }

    list.forEach(item => {
      if (item.pagePath.indexOf('/') !== 0) {
        item.pagePath = '/' + item.pagePath
      }
    })

    this.state = {
      list,
      selectedIndex: -1,
      status: STATUS_SHOW
    }
  }

  homePage = ''
  tabbar = null
  tabbarPos = 'bottom'

  getCurrentUrl () {
    const currentPagename = this.props.currentPagename
    const routerMode = this.props.conf.mode
    const routerBasename = this.props.conf.basename || '/'
    let url
    if (routerMode === 'hash') {
      const href = window.location.href
      const hashIndex = href.indexOf('#')
      url = hashIndex === -1
        ? ''
        : href.substring(hashIndex + 1)
    } else if (routerMode === 'multi') {
      url = currentPagename
    } else {
      url = location.pathname
    }
    const processedUrl = addLeadingSlash(stripBasename(url, routerBasename))
    return processedUrl === '/'
      ? this.homePage
      : processedUrl
  }

  getOriginUrl = url => {
    const customRoute = this.customRoutes.filter(([originUrl, customUrl]) => {
      const patha = splitUrl(customUrl).path
      const pathb = splitUrl(url).path
      return patha === pathb
    })
    return customRoute.length ? customRoute[0][0] : url
  }

  getSelectedIndex = url => {
    let foundIndex = -1
    this.state.list.forEach(({ pagePath }, idx) => {
      const patha = splitUrl(url).path
      const pathb = splitUrl(pagePath).path
      if (patha === pathb) {
        foundIndex = idx
      }
    })
    return foundIndex
  }

  switchTab = (index) => {
    this.setState({
      selectedIndex: index
    })
    Taro.redirectTo && Taro.redirectTo({
      url: this.state.list[index].pagePath
    })
  }

  tabbarRef = (ref) => {
    const domNode = findDOMNode(ref)
    this.tabbar = domNode
  }

  switchTabHandler = ({ url, successHandler, errorHandler }) => {
    const routerMode = this.props.conf.mode
    const currentUrl = routerMode === 'multi'
      ? this.props.currentPagename
      : this.getOriginUrl(this.getCurrentUrl() || this.homePage)
    const nextTab = resolvePathname(url, currentUrl)
    const foundIndex = this.getSelectedIndex(nextTab)

    if (foundIndex > -1) {
      this.switchTab(foundIndex)
      successHandler({
        errMsg: `switchTab:ok`
      })
    } else {
      errorHandler({
        errMsg: `switchTab:fail page "${nextTab}" is not found`
      })
    }
  }

  routerChangeHandler = ({ toLocation } = {}) => {
    let currentPage

    if (toLocation && toLocation.path) {
      const tmpPath = addLeadingSlash(toLocation.path)
      currentPage = tmpPath === '/'
        ? this.homePage
        : tmpPath
    } else {
      currentPage = this.getCurrentUrl()
    }

    this.setState({
      selectedIndex: this.getSelectedIndex(this.getOriginUrl(currentPage))
    })
  }

  setTabBarBadgeHandler = ({ index, text, successHandler, errorHandler }) => {
    const list = this.state.list
    if (index in list) {
      list[index].showRedDot = false
      list[index].badgeText = text
      this.setState({}, () => {
        successHandler({
          errMsg: 'setTabBarBadge:ok'
        })
      })
    } else {
      errorHandler({
        errMsg: `setTabBarBadge:fail tabbar item not found`
      })
    }
  }

  removeTabBarBadgeHandler = ({ index, successHandler, errorHandler }) => {
    const list = this.state.list
    if (index in list) {
      list[index].badgeText = null
      this.setState({}, () => {
        successHandler({
          errMsg: 'removeTabBarBadge:ok'
        })
      })
    } else {
      errorHandler({
        errMsg: `removeTabBarBadge:fail tabbar item not found`
      })
    }
  }

  showTabBarRedDotHandler = ({ index, successHandler, errorHandler }) => {
    const list = this.state.list
    if (index in list) {
      list[index].badgeText = null
      list[index].showRedDot = true
      this.setState({}, () => {
        successHandler({
          errMsg: 'showTabBarRedDot:ok'
        })
      })
    } else {
      errorHandler({
        errMsg: `showTabBarRedDot:fail tabbar item not found`
      })
    }
  }

  hideTabBarRedDotHandler = ({ index, successHandler, errorHandler }) => {
    const list = this.state.list
    if (index in list) {
      list[index].showRedDot = false
      this.setState({}, () => {
        successHandler({
          errMsg: 'hideTabBarRedDot:ok'
        })
      })
    } else {
      errorHandler({
        errMsg: `hideTabBarRedDot:fail tabbar item not found`
      })
    }
  }

  showTabBarHandler = ({ successHandler }) => {
    this.setState({
      status: STATUS_SHOW
    }, () => {
      successHandler({
        errMsg: 'showTabBar:ok'
      })
    })
  }

  hideTabBarHandler = ({ animation, successHandler }) => {
    this.setState({
      status: animation ? STATUS_SLIDEOUT : STATUS_HIDE
    }, () => {
      successHandler({
        errMsg: 'hideTabBar:ok'
      })
    })
  }

  bindEvent () {
    Taro.eventCenter.on('__taroRouterChange', this.routerChangeHandler)
    Taro.eventCenter.on('__taroSwitchTab', this.switchTabHandler)
    Taro.eventCenter.on('__taroSetTabBarBadge', this.setTabBarBadgeHandler)
    Taro.eventCenter.on('__taroRemoveTabBarBadge', this.removeTabBarBadgeHandler)
    Taro.eventCenter.on('__taroShowTabBarRedDotHandler', this.showTabBarRedDotHandler)
    Taro.eventCenter.on('__taroHideTabBarRedDotHandler', this.hideTabBarRedDotHandler)
    Taro.eventCenter.on('__taroShowTabBar', this.showTabBarHandler)
    Taro.eventCenter.on('__taroHideTabBar', this.hideTabBarHandler)
  }

  removeEvent () {
    Taro.eventCenter.off('__taroRouterChange', this.routerChangeHandler)
    Taro.eventCenter.off('__taroSwitchTab', this.switchTabHandler)
    Taro.eventCenter.off('__taroSetTabBarBadge', this.setTabBarBadgeHandler)
    Taro.eventCenter.off('__taroRemoveTabBarBadge', this.removeTabBarBadgeHandler)
    Taro.eventCenter.off('__taroShowTabBarRedDotHandler', this.showTabBarRedDotHandler)
    Taro.eventCenter.off('__taroHideTabBarRedDotHandler', this.hideTabBarRedDotHandler)
    Taro.eventCenter.off('__taroShowTabBarHandler', this.showTabBarHandler)
    Taro.eventCenter.off('__taroHideTabBarHandler', this.hideTabBarHandler)
  }

  componentDidMount () {
    this.tabbarPos = this.vnode.dom.nextElementSibling
      ? 'top'
      : 'bottom'
    this.bindEvent()
    this.routerChangeHandler()
  }

  componentWillUnmount () {
    this.removeEvent()
  }

  render () {
    const { conf, tabbarPos = 'bottom' } = this.props
    const { status } = this.state
    const containerCls = classNames('weui-tabbar', {
      [`taro-tabbar__border-${conf.borderStyle || 'black'}`]: true
    })
    const shouldHideTabBar = this.state.selectedIndex === -1 || status === STATUS_HIDE
    const shouldSlideout = status === STATUS_SLIDEOUT

    return (
      <div
        ref={this.tabbarRef}
        className={classNames(
          basicTabBarClassName,
          `${basicTabBarClassName}-${tabbarPos}`, {
            [hideTabBarClassName]: shouldHideTabBar,
            [hideTabBarWithAnimationClassName]: shouldSlideout
          })} >
        <div
          className={containerCls}
          style={{
            backgroundColor: conf.backgroundColor || ''
          }}
        >
          {this.state.list.map((item, index) => {
            const isSelected = this.state.selectedIndex === index
            let textColor
            let iconPath
            if (isSelected) {
              textColor = conf.selectedColor
              iconPath = item.selectedIconPath
            } else {
              textColor = conf.color || ''
              iconPath = item.iconPath
            }
            return (
              <TabbarItem
                index={index}
                onSelect={this.switchTab.bind(this)}
                isSelected={isSelected}
                textColor={textColor}
                iconPath={iconPath}
                text={item.text}
                badgeText={item.badgeText}
                showRedDot={item.showRedDot} />
            )
          })}
        </div>
      </div>
    )
  }
}

export default Tabbar
