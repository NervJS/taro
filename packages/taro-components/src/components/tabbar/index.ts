export * from './tabbar-item'
export * from './tabbar'
