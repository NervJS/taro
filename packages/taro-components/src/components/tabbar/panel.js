import 'weui'
import Nerv from 'nervjs'
import classNames from 'classnames'
import View from '../view'
export default class TabbarPanel extends Nerv.Component {
  render () {
    const { children, className, ...reset } = this.props
    const cls = classNames('taro-tabbar__panel', className)
    return (
      <View className={cls} {...reset}>
        {children}
      </View>
    )
  }
}
