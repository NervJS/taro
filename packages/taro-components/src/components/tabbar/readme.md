# taro-tabbar



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type   | Default     |
| -------- | --------- | ----------- | ------ | ----------- |
| `conf`   | --        |             | `Conf` | `undefined` |


## Events

| Event       | Description | Type               |
| ----------- | ----------- | ------------------ |
| `longpress` |             | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
