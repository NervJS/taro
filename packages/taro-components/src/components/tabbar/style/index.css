html, body {
  height: 100%;
}

#app {
  height: 100%;
}

.taro-tabbar__border-white::before {
  border-top-color: #fff !important;
}

.taro-tabbar__container {
  display: flex;
  height: 100%;
  flex-direction: column;
  overflow: hidden;
}

.taro-tabbar__panel {
  flex: 1;
  position: relative;
  overflow: auto;
  -webkit-overflow-scrolling: touch;
}

.taro-tabbar__tabbar {
  position: relative;
  height: 50px;
  width: 100%;
  transition: bottom .2s, top .2s;
}

.taro-tabbar__tabbar-top {
  top: 0;
}

.taro-tabbar__tabbar-bottom {
  bottom: 0;
}

.taro-tabbar__tabbar-hide {
  display: none;
}

.taro-tabbar__tabbar-slideout {
  top: -52px;
  flex: 0 0;
}

.taro-tabbar__panel + .taro-tabbar__tabbar-slideout {
  top: auto;
  bottom: -52px;
}
