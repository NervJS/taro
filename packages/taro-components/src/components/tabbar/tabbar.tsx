import { Component, Prop, h, ComponentInterface, Host, State, Event, EventEmitter, Element } from '@stencil/core'
import classNames from 'classnames'
import resolvePathname from 'resolve-pathname'
import { splitUrl } from '../../utils'
import { TabbarItem } from './tabbar-item'
const Taro = require('@tarojs/taro')

// const removeLeadingSlash = str => str.replace(/^\.?\//, '')
// const removeTrailingSearch = str => str.replace(/\?[\s\S]*$/, '')
const addLeadingSlash = str => str[0] === '/' ? str : `/${str}`

const hasBasename = (path, prefix) =>
  new RegExp('^' + prefix + '(\\/|\\?|#|$)', 'i').test(path)

const stripBasename = (path, prefix) =>
  hasBasename(path, prefix) ? path.substr(prefix.length) : path

const STATUS_SHOW = 0
const STATUS_HIDE = 1
const STATUS_SLIDEOUT = 2

const basicTabBarClassName = 'taro-tabbar__tabbar'
const hideTabBarClassName = 'taro-tabbar__tabbar-hide'
const hideTabBarWithAnimationClassName = 'taro-tabbar__tabbar-slideout'

interface RouterHandler {
  index: string
  text: string
  url: string
  successHandler: Function
  errorHandler: Function
  animation?: boolean
}

export interface Conf {
  color: string
  selectedColor: string
  backgroundColor: string
  borderStyle?: 'black' | 'white'
  list: TabbarList[]
  position?: 'bottom' | 'top'
  custom: boolean
  customRoutes: Record<string, string>
  mode: 'hash' | 'browser'
  basename: string
  homePage: string
  currentPagename: string
}

interface TabbarList {
  pagePath: string
  text: string
  iconPath?: string
  selectedIconPath?: string
  badgeText?: string
  showRedDot?: boolean
}

@Component({
  tag: 'taro-tabbar',
  styleUrl: './style/index.scss'
})
export class Tabbar implements ComponentInterface {
  private homePage = ''

  private customRoutes: Array<string[]> = []

  private tabbarPos: 'top' | 'bottom' = 'bottom'

  @Prop() conf: Conf

  @State() list: TabbarList[]

  @State() selectedIndex = -1

  @State() status: 0 | 1 | 2 = STATUS_SHOW

  @Event({
    eventName: 'longpress'
  }) onLongPress: EventEmitter

  @Element() tabbar: HTMLDivElement

  constructor () {
    const list = this.conf.list
    const customRoutes = this.conf.customRoutes
    if (
      Object.prototype.toString.call(list) !== '[object Array]' ||
      list.length < 2 ||
      list.length > 5
    ) {
      throw new Error('tabBar 配置错误')
    }

    this.homePage = addLeadingSlash(this.conf.homePage)
    for (const key in customRoutes) {
      this.customRoutes.push([key, customRoutes[key]])
    }

    list.forEach(item => {
      if (item.pagePath.indexOf('/') !== 0) {
        item.pagePath = '/' + item.pagePath
      }
    })

    this.list = list
  }

  getCurrentUrl () {
    const routerMode = this.conf.mode
    const routerBasename = this.conf.basename || '/'
    let url
    if (routerMode === 'hash') {
      const href = window.location.href
      const hashIndex = href.indexOf('#')
      url = hashIndex === -1
        ? ''
        : href.substring(hashIndex + 1)
    } else {
      url = location.pathname
    }
    const processedUrl = addLeadingSlash(stripBasename(url, routerBasename))
    return processedUrl === '/' ? this.homePage : processedUrl
  }

  getOriginUrl = (url: string) => {
    const customRoute = this.customRoutes.filter(([, customUrl]) => {
      const patha = splitUrl(customUrl).path
      const pathb = splitUrl(url).path
      return patha === pathb
    })
    return customRoute.length ? customRoute[0][0] : url
  }

  getSelectedIndex = (url: string) => {
    let foundIndex = -1
    this.list.forEach(({ pagePath }, idx) => {
      const patha = splitUrl(url).path
      const pathb = splitUrl(pagePath).path
      if (patha === pathb) {
        foundIndex = idx
      }
    })
    return foundIndex
  }

  switchTab = (index: number) => {
    this.selectedIndex = index
    Taro.navigateTo({
      url: this.list[index].pagePath
    })
  }

  switchTabHandler = ({ url, successHandler, errorHandler }: RouterHandler) => {
    const currentUrl = this.getOriginUrl(this.getCurrentUrl() || this.homePage)
    const nextTab = resolvePathname(url, currentUrl)
    const foundIndex = this.getSelectedIndex(nextTab)

    if (foundIndex > -1) {
      this.switchTab(foundIndex)
      successHandler({
        errMsg: 'switchTab:ok'
      })
    } else {
      errorHandler({
        errMsg: `switchTab:fail page "${nextTab}" is not found`
      })
    }
  }

  routerChangeHandler = (options?) => {
    let toLocation
    let currentPage

    if (options) {
      toLocation = options.toLocation
    }

    if (toLocation && toLocation.path) {
      const tmpPath = addLeadingSlash(toLocation.path)
      currentPage = tmpPath === '/'
        ? this.homePage
        : tmpPath
    } else {
      currentPage = this.getCurrentUrl()
    }

    this.selectedIndex = this.getSelectedIndex(this.getOriginUrl(currentPage))
  }

  setTabBarBadgeHandler = ({ index, text, errorHandler }: RouterHandler) => {
    const list = this.list
    if (index in list) {
      this.list[index].showRedDot = false
      this.list[index].badgeText = text
    } else {
      errorHandler({
        errMsg: 'setTabBarBadge:fail tabbar item not found'
      })
    }
  }

  removeTabBarBadgeHandler = ({ index, successHandler, errorHandler }: RouterHandler) => {
    const list = this.list
    if (index in list) {
      this.list[index].badgeText = null
      this.list[index].badgeText = null
      successHandler({
        errMsg: 'removeTabBarBadge:ok'
      })
    } else {
      errorHandler({
        errMsg: 'removeTabBarBadge:fail tabbar item not found'
      })
    }
  }

  showTabBarRedDotHandler = ({ index, successHandler, errorHandler }: RouterHandler) => {
    const list = this.list
    if (index in list) {
      list[index].badgeText = null
      list[index].showRedDot = true
      successHandler({
        errMsg: 'showTabBarRedDot:ok'
      })
    } else {
      errorHandler({
        errMsg: 'showTabBarRedDot:fail tabbar item not found'
      })
    }
  }

  hideTabBarRedDotHandler = ({ index, successHandler, errorHandler }: RouterHandler) => {
    const list = this.list
    if (index in list) {
      list[index].showRedDot = false
      successHandler({
        errMsg: 'hideTabBarRedDot:ok'
      })
    } else {
      errorHandler({
        errMsg: 'hideTabBarRedDot:fail tabbar item not found'
      })
    }
  }

  showTabBarHandler = ({ successHandler }) => {
    this.status = STATUS_SHOW
    successHandler({
      errMsg: 'showTabBar:ok'
    })
  }

  hideTabBarHandler = ({ animation, successHandler }) => {
    this.status = animation ? STATUS_SLIDEOUT : STATUS_HIDE
    successHandler({
      errMsg: 'hideTabBar:ok'
    })
  }

  bindEvent () {
    Taro.eventCenter.on('__taroRouterChange', this.routerChangeHandler)
    Taro.eventCenter.on('__taroSwitchTab', this.switchTabHandler)
    Taro.eventCenter.on('__taroSetTabBarBadge', this.setTabBarBadgeHandler)
    Taro.eventCenter.on('__taroRemoveTabBarBadge', this.removeTabBarBadgeHandler)
    Taro.eventCenter.on('__taroShowTabBarRedDotHandler', this.showTabBarRedDotHandler)
    Taro.eventCenter.on('__taroHideTabBarRedDotHandler', this.hideTabBarRedDotHandler)
    Taro.eventCenter.on('__taroShowTabBar', this.showTabBarHandler)
    Taro.eventCenter.on('__taroHideTabBar', this.hideTabBarHandler)
  }

  removeEvent () {
    Taro.eventCenter.off('__taroRouterChange', this.routerChangeHandler)
    Taro.eventCenter.off('__taroSwitchTab', this.switchTabHandler)
    Taro.eventCenter.off('__taroSetTabBarBadge', this.setTabBarBadgeHandler)
    Taro.eventCenter.off('__taroRemoveTabBarBadge', this.removeTabBarBadgeHandler)
    Taro.eventCenter.off('__taroShowTabBarRedDotHandler', this.showTabBarRedDotHandler)
    Taro.eventCenter.off('__taroHideTabBarRedDotHandler', this.hideTabBarRedDotHandler)
    Taro.eventCenter.off('__taroShowTabBarHandler', this.showTabBarHandler)
    Taro.eventCenter.off('__taroHideTabBarHandler', this.hideTabBarHandler)
  }

  componentDidLoad () {
    this.tabbarPos = this.tabbar.nextElementSibling ? 'top' : 'bottom'
    this.bindEvent()
    this.routerChangeHandler()
  }

  componentDidUnload () {
    this.removeEvent()
  }

  render () {
    const { conf, tabbarPos = 'bottom' } = this
    const status = this.status
    const containerCls = classNames('weui-tabbar', {
      [`taro-tabbar__border-${conf.borderStyle || 'black'}`]: true
    })
    const shouldHideTabBar = this.selectedIndex === -1 || status === STATUS_HIDE
    const shouldSlideout = status === STATUS_SLIDEOUT

    return (
      <Host
        class={classNames(
          basicTabBarClassName,
          `${basicTabBarClassName}-${tabbarPos}`, {
            [hideTabBarClassName]: shouldHideTabBar,
            [hideTabBarWithAnimationClassName]: shouldSlideout
          })} >
        <div
          class={containerCls}
          style={{
            backgroundColor: conf.backgroundColor || ''
          }}
        >
          {this.list.map((item, index) => {
            const isSelected = this.selectedIndex === index
            let textColor
            let iconPath
            if (isSelected) {
              textColor = conf.selectedColor
              iconPath = item.selectedIconPath
            } else {
              textColor = conf.color || ''
              iconPath = item.iconPath
            }
            return (
              <TabbarItem
                index={index}
                onSelect={this.switchTab.bind(this)}
                isSelected={isSelected}
                textColor={textColor}
                iconPath={iconPath}
                text={item.text}
                badgeText={item.badgeText}
                showRedDot={item.showRedDot} />
            )
          })}
        </div>
      </Host>
    )
  }
}
