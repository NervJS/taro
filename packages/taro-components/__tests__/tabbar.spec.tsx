import { h } from '@stencil/core'
import { newSpecPage, SpecPage } from '@stencil/core/testing'

import { Tabbar } from '../src/components/tabbar/tabbar'

describe('Tabbar', () => {
  let page: SpecPage

  it('base', async () => {
    page = await newSpecPage({
      components: [Tabbar],
      template: () => (<taro-tabbar conf={{
        list: [
          {
            pagePath: 'pages/index/index',
            text: '首页'
          },
          {
            pagePath: 'pages/cate/index',
            text: '分类'
          },
          {
            pagePath: 'pages/cart/index',
            text: '购物车'
          },
          {
            pagePath: 'pages/my/index',
            text: '个人中心'
          }
        ]
      }} />),
    })
    expect(page.root).toMatchSnapshot()
  })
})
