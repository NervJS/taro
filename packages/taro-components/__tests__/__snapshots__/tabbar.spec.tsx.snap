// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabbar base 1`] = `
<taro-tabbar class="taro-tabbar__tabbar taro-tabbar__tabbar-bottom taro-tabbar__tabbar-hide">
  <div class="taro-tabbar__border-black weui-tabbar" style="height: inherit;">
    <a class="weui-tabbar__item" href="javascript:;">
      <span style="display: inline-block; position: relative;">
        <img alt="" class="weui-tabbar__icon">
      </span>
      <p class="weui-tabbar__label">
        首页
      </p>
    </a>
    <a class="weui-tabbar__item" href="javascript:;">
      <span style="display: inline-block; position: relative;">
        <img alt="" class="weui-tabbar__icon">
      </span>
      <p class="weui-tabbar__label">
        分类
      </p>
    </a>
    <a class="weui-tabbar__item" href="javascript:;">
      <span style="display: inline-block; position: relative;">
        <img alt="" class="weui-tabbar__icon">
      </span>
      <p class="weui-tabbar__label">
        购物车
      </p>
    </a>
    <a class="weui-tabbar__item" href="javascript:;">
      <span style="display: inline-block; position: relative;">
        <img alt="" class="weui-tabbar__icon">
      </span>
      <p class="weui-tabbar__label">
        个人中心
      </p>
    </a>
  </div>
</taro-tabbar>
`;
